### PR TITLE
fix: add validation to datasource create flow and error handling

### LINKED ISSUE #25200: [Bug]: Error handlign for when action executed against non-existent datasource storage
### Is there an existing issue for this?
- [x] I have searched the existing issues

### Description
"When app has only production env datasource configured, in staging env for db, on running these queries it gives internal server error"

### Steps To Reproduce
"When app has only production env datasource configured, in staging env for db, on running these queries it gives internal server error"

### Public Sample App
_No response_

### Environment
Deploy Preview

### Issue video log
_No response_

### Version
Cloud

### PR DESCRIPTION
**Pull request by sondermanish**
#### PR fixes following issue(s)
Fixes #25200, #25204  

#### Type of change
- Bug fix (non-breaking change which fixes an issue)

#### How Has This Been Tested?
- [ ] Manual
- [ ] Jest
- [ ] Cypress

#### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

File: app/server/appsmith-server/src/test/java/com/appsmith/server/services/DatasourceStorageServiceTest.java [hunk 1]
```diff
@@ -81,4 +97,78 @@ public void verifyFindByDatasourceId() {
...
+        datasourceStorageService.create(datasourceStorage).block();
+        StepVerifier.create(datasourceStorageService.create(datasourceStorage)).verifyErrorSatisfies(error -> {
+            assertThat(error).isInstanceOf(AppsmithException.class);
+            assertThat(((AppsmithException) error).getAppErrorCode()).isEqualTo("AE-APP-4091");
+        });
+    }
+
+    @Test
+    @WithUserDetails(value = "api_user")
+    public void verifyStorageCreationSuccedsWithDifferentEnvironmentId() {
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/test/java/com/appsmith/server/services/DatasourceStorageServiceTest.java:138]: Nit: Typo in method name

File: app/server/appsmith-server/src/test/java/com/appsmith/server/services/DatasourceStorageServiceTest.java [hunk 2]
```diff
@@ -25,6 +34,7 @@
 import static org.assertj.core.api.Assertions.assertThat;
 
 @ExtendWith(SpringExtension.class)
+@DirtiesContext
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/test/java/com/appsmith/server/services/DatasourceStorageServiceTest.java:37]: Why is this dirtying the context?

File: app/server/appsmith-server/src/test/java/com/appsmith/server/solutions/ce/ActionExecutionSolutionCETest.java [hunk 3]
```diff
@@ -1852,4 +1850,51 @@ public void executeAction_actionOnMockDatasource_success() {
...
+                    executeActionDTO.setViewMode(false);
+                    return actionExecutionSolution.executeAction(executeActionDTO, defaultEnvironmentId);
+                });
+
+        Mockito.doReturn(Mono.empty()).when(datasourceStorageService).findByDatasourceAndEnvironmentId(any(), any());
+
+        StepVerifier.create(resultMono).verifyErrorSatisfies(error -> {
+            assertThat(error).isInstanceOf(AppsmithException.class);
+
+            assertThat(((AppsmithException) error).getAppErrorCode()).isEqualTo("AE-DTS-4011");
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/test/java/com/appsmith/server/solutions/ce/ActionExecutionSolutionCETest.java:1897]: Let's use the ref to app error code? Static strings have a way of getting outdated

File: app/server/appsmith-server/src/test/java/com/appsmith/server/services/DatasourceStorageServiceTest.java (hunk 2 above)
Comment (sondermanish) [reply app/server/appsmith-server/src/test/java/com/appsmith/server/services/DatasourceStorageServiceTest.java:37]: The tests are interlinked in the way of execution, After adding validation if we try to create storage with same datasourceId and environmentId we are getting error

**Comment by sondermanish** [comment]
Skipping test plan as it's covered in the EE PR.

**Pull request merged**